src/org/traccar/protocol/NavisProtocolDecoder.java

```diff
@@ -272,41 +317,343 @@ private Object processArray(DeviceSession deviceSession, Channel channel, ByteBu
+
+            position.set(Position.KEY_EVENT, buf.readUnsignedShortLE());
+            buf.readUnsignedInt(); // event time
+
+            int navSensorState = buf.readUnsignedByte();
+            position.setValid(BitUtil.check(navSensorState, 1));
+            position.set(Position.KEY_SATELLITES, BitUtil.from(navSensorState, 2));
+
+            position.setTime(new DateBuilder(new Date(buf.readUnsignedIntLE() * 1000)).getDate());
+            position.setLatitude(buf.readIntLE() / 600000.0);
+            position.setLongitude(buf.readIntLE() / 600000.0);
+            position.setAltitude(buf.readIntLE() * 0.1);
+            position.setSpeed(UnitsConverter.knotsFromKph(buf.readFloatLE()));
+            position.setCourse(buf.readUnsignedShortLE());
+            position.set(Position.KEY_ODOMETER, buf.readFloatLE() * 1000);
+
+            buf.skipBytes(length - buf.readerIndex() - 1); // skip unused part
         }
+
+        return new ParseResult(index, position);
+    }
+
+    private interface FlexPositionParser {
+        ParseResult parsePosition(DeviceSession deviceSession, ByteBuf buf);
+    }
+
+    private Object processFlexSingle(
+            FlexPositionParser parser, String flexHeader, DeviceSession deviceSession, Channel channel, ByteBuf buf) {
+        if (!flexHeader.equals("~C")) {
+            buf.readUnsignedInt(); // event index
+        }
+
+        ParseResult result = parser.parsePosition(deviceSession, buf);
+
+        ByteBuf response = Unpooled.buffer(6);
+        response.writeCharSequence(flexHeader, StandardCharsets.US_ASCII);
+        response.writeIntLE((int) result.getId());
+        sendFlexReply(channel, response);
+
+        if (result.getPosition().getFixTime() == null) {
+            return null;
+        }
+
+        return result.getPosition();
+    }
+
+    private Object processFlexArray(
+            FlexPositionParser parser, String flexHeader, DeviceSession deviceSession, Channel channel, ByteBuf buf) {
+        List<Position> positions = new LinkedList<>();
+        int count = buf.readUnsignedByte();
+
+        for (int i = 0; i < count; i++) {
+            Position position = parser.parsePosition(deviceSession, buf).getPosition();
+            if (position.getFixTime() != null) {
+                positions.add(position);
+            }
+        }
+
+        ByteBuf response = Unpooled.buffer(6);
+        response.writeCharSequence(flexHeader, StandardCharsets.US_ASCII);
+        response.writeByte(count);
+        sendFlexReply(channel, response);
+
+        if (positions.isEmpty()) {
+            return null;
+        }
+
+        return positions;
+    }
+
+    private Object processFlexNegotiation(Channel channel, ByteBuf buf) {
+        if ((byte) buf.readUnsignedByte() != (byte) 0xB0) {
+            return null;
+        }
+
+        byte flexProtocolVersion, flexStructVersion;
+
+        flexProtocolVersion = (byte) buf.readUnsignedByte();
```

Comment:
Why do we need byte conversion everywhere? I think it can be removed.

Reply:
Fixed